standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/retryhandling/RetryHandler.java

```diff
@@ -0,0 +1,305 @@
+   * @param function The {@link TransactionalFunction} to execute.
+   * @return Returns with the result of the execution of the passed {@link TransactionalFunction}.
+   * @param <Result> Type of the result
+   * @throws MetaException Thrown in case of execution error.
+   */
+  public <Result> Result executeWithRetry(RetryCallProperties properties, TransactionalFunction<Result> function) 
+      throws MetaException {
+    return executeWithRetryInternal(properties, function, retryLimit, ALLOWED_REPEATED_DEADLOCKS);
+  }
+
+  /**
+   * Executes the passed {@link TransactionalFunction}, without retry in case of failure.
+   * @param properties {@link RetryCallProperties} instance used to fine-tune/alter the execution-mechanism.
+   * @param function The {@link TransactionalFunction} to execute.
+   * @return Returns with the result of the execution of the passed {@link TransactionalFunction}.
+   * @param <Result> Type of the result
+   * @throws MetaException Thrown in case of execution error.
+   */
+  public <Result> Result executeWithoutRetry(RetryCallProperties properties, TransactionalFunction<Result> function) 
+      throws MetaException {
+    return executeWithRetryInternal(properties, function, 0, 0);
+  }
+
+  private <Result> Result executeWithRetryInternal(RetryCallProperties properties, TransactionalFunction<Result> function
+      , int retryCount, int deadlockCount) throws MetaException {
+    Objects.requireNonNull(function, "RetryFunction<Result> cannot be null!");
+    Objects.requireNonNull(properties, "RetryCallProperties cannot be null!");
+
+    if (jdbcTemplate.hasRetryContext() && properties.getRetryPropagation().canJoinContext()) {
+      /*
+        If there is a RetryState in the ThreadLocal and we are allowed to join it, we can skip establishing a nested 
+        retry-call (and transaction).
+      */
+      LOG.info("Already established retry-context detected, current retry-call will join it. The passed CallProperties " +
+          "instance will be ignored, using the original one.");
+      try {
+        return function.execute(jdbcTemplate);
+      } catch (SQLException e) {
+        throw new UncategorizedSQLException(null, null, e);
+      }
+    }
+
+    if (!properties.getRetryPropagation().canCreateContext()) {
+      throw new MetaException("The current RetryPropagation mode (" + properties.getRetryPropagation().name() +
+          ") allows only to join an existing retry-context, but there is no context to join!");
+    }
+
+    LOG.debug("Running retry function:" + properties);
+
+    if (org.apache.commons.lang3.StringUtils.isEmpty(properties.getCaller())) {
+      properties.withCallerId(function.toString());
+    }
+
+    try {
+      if (properties.isLockInternally()) {
+        lockInternal();
+      }
+      try {
+        DefaultTransactionDefinition transactionDefinition = new DefaultTransactionDefinition(TransactionDefinition.PROPAGATION_REQUIRED);
```

Comment:
Can't we have @Transaction annotation similar to Retry on method level where we could define the propagation?

Reply:
Created the propagation in RetryCallProperties. The idea was to tie together the retry and transaction handling: In case of failure rollback AND retry. In case of success commit AND do no retry. Handling them independently would allow more fine grained control, but would also introduce extra complexity: cases like doing a retry with the old connection and txn, or committing the txn, and still doing a retry. I'm not sure it's worth allowing these, as currently I cannot see places where such cases are necessary.